classifier/bisac.py

```diff
@@ -0,0 +1,605 @@
+                must_match, remaining_subject
+            )
+
+        if match_so_far:
+            # Everything that had to match, did.
+            self.caught.append(subject)
+            return self.result
+
+        # Something that had to match, didn't.
+        return None
+
+    def _consume(self, rules, subject):
+        """The first token (and possibly more) of the rules must match the
+        first token (and possibly more) of the subject.
+
+        All matched rule and subject tokens are consumed.
+
+        :return: A 3-tuple (could_match, new_rules, new_subject)
+
+        could_match is a boolean that is False if we now know that the
+        subject does not match the rule, and True if it might still
+        match the rule.
+
+        new_rules contains the tokens in the ruleset that have yet to
+        be activated.
+        
+        new_subject contains the tokens in the subject that have yet
+        to be checked.
+        """
+        if not rules:
+            # An empty ruleset matches everything.
+            return True, rules, subject
+
+        if not subject and rules != [anything]:
+            # Apart from [anything], no non-empty ruleset matches an
+            # empty subject.
+            return False, rules, subject
+
+        # Figure out which rule we'll be applying. We won't need it
+        # again, so we can remove it from the ruleset.
+        rule_token = rules.pop(0)
+        if rule_token == anything:
+            # This is the complicated one.
+
+            if not rules:
+                # If the final rule is 'anything', then that's redundant,
+                # but we can declare success and stop.
+                return True, rules, subject
+
+            # At this point we know that 'anything' is followed by some
+            # other rule token.
+            next_rule = rules.pop(0)
+
+            # We can consume as many subject tokens as necessary, but
+            # eventually a subject token must match this subsequent
+            # rule token.
+            while subject:
+                subject_token = subject.pop(0)
+                submatch, ignore1, ignore2 = self._consume(
+                    [next_rule], [subject_token]
+                )
+                if submatch:
+                    # We had to remove some number of subject tokens,
+                    # but we found one that matches the next rule.
+                    return True, rules, subject
+                elif not subject:
+                    # We are out of subject tokens, and we never found a
+                    # candidate for the next rule token.
+                    return False, rules, subject
+                else:
+                    # That token didn't match, but maybe the next one will.
+                    pass
+
+            # We went through the entire remaining subject and didn't
+            # find a match for the rule token that follows 'anything'.
+            return False, must_match, subject
```

Comment:
I don't think this code is reachable, but if it was reached, `must_match` would be undefined.